test/provisioner/provision_aws.go

```diff
@@ -0,0 +1,379 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package provisioner
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsConfig "github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/aws/aws-sdk-go-v2/service/iam"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/e2e-framework/pkg/envconf"
+)
+
+func init() {
+	// Add this implementation to the list of provisioners.
+	newProvisionerFunctions["aws"] = NewAWSProvisioner
+}
+
+// S3Bucket represents an S3 bucket where the podvm image should be uploaded
+type S3Bucket struct {
+	Client *s3.Client
+	Name   string // Bucket name
+	Key    string // Object key
+}
+
+// AMIImage represents an AMI image
+type AMIImage struct {
+	Client          *ec2.Client
+	Description     string // Image description
+	DiskDescription string // Disk description
+	DiskFormat      string // Disk format
+	EBSSnapshotId   string // EBS disk snapshot ID
+	ID              string // AMI image ID
+	RootDeviceName  string // Root device name
+}
+
+// AWSProvisioner implements the CloudProvision interface.
+type AWSProvisioner struct {
+	AwsConfig aws.Config
+	iamClient *iam.Client
+	ec2Client *ec2.Client
+	s3Client  *s3.Client
+	Bucket    *S3Bucket
+	Image     *AMIImage
+}
+
+// NewAWSProvisioner instantiates the AWS provisioner
+func NewAWSProvisioner(properties map[string]string) (CloudProvisioner, error) {
+	cfg, err := awsConfig.LoadDefaultConfig(context.TODO())
+	if err != nil {
+		log.Error("Failed to load AWS config")
+		return nil, err
+	}
+
+	if properties["aws_region"] != "" {
+		cfg.Region = properties["aws_region"]
+	}
+
+	ec2Client := ec2.NewFromConfig(cfg)
+	return &AWSProvisioner{
+		AwsConfig: cfg,
+		iamClient: iam.NewFromConfig(cfg),
+		ec2Client: ec2.NewFromConfig(cfg),
+		s3Client:  s3.NewFromConfig(cfg),
+		Bucket: &S3Bucket{
+			Client: s3.NewFromConfig(cfg),
+			Name:   "peer-pods-tests",
+			Key:    "", // To be defined when the file is uploaded
+		},
+		Image: NewAMIImage(ec2Client, properties),
+	}, nil
+}
+
+func (aws *AWSProvisioner) CreateCluster(ctx context.Context, cfg *envconf.Config) error {
+	return nil
+}
+
+func (aws *AWSProvisioner) CreateVPC(ctx context.Context, cfg *envconf.Config) error {
+	return nil
+}
+
+func (aws *AWSProvisioner) DeleteCluster(ctx context.Context, cfg *envconf.Config) error {
+	return nil
+}
+
+func (aws *AWSProvisioner) DeleteVPC(ctx context.Context, cfg *envconf.Config) error {
+	return nil
+}
+
+func (aws *AWSProvisioner) GetProperties(ctx context.Context, cfg *envconf.Config) map[string]string {
+	return make(map[string]string)
+}
+
+func (a *AWSProvisioner) UploadPodvm(imagePath string, ctx context.Context, cfg *envconf.Config) error {
+	// AWS EC2 image-import does not support qcow2 files so convert the image to raw format.
```

Comment:
have you consider using also the direct packer for AWS image creation (aws/image)? it's much than uploading an image which i suppose would be better for CI purposes. (and be done later of course...)

Reply:
I didn't. Could you send me pointers so I can consider to implement in follow up changes?

The upload process here basically mimics the instructions in REAME (./aws/README.md) but implemented in Go: convert the qcow2 to raw disk (function `ConvertQcow2ToRaw()`) then upload (`raw-to-ami.sh`).

Reply:
https://github.com/confidential-containers/cloud-api-adaptor/tree/main/aws#build-pod-vm-image

Reply:
it's the [option 1 ](https://github.com/confidential-containers/cloud-api-adaptor/tree/main/aws#option-1-modifying-existing-marketplace-image)in the same README